src/qutip_qip/algorithms/qpe.py

```diff
@@ -0,0 +1,212 @@
+import numpy as np
+from qutip import Qobj
+from qutip_qip.operations import Gate, ControlledGate
+from qutip_qip.circuit import QubitCircuit
+from qutip_qip.decompose import decompose_one_qubit_gate
+
+__all__ = ["qpe", "CustomGate"]
+
+
+class CustomGate(Gate):
+    """
+    Custom gate that wraps an arbitrary quantum operator.
+    """
+    def __init__(self, targets, U, **kwargs):
+        super().__init__(targets=targets, **kwargs)
+        self.targets = targets if isinstance(targets, list) else [targets]
+        self.U = U  
+        self.kwargs = kwargs
+        self.latex_str = r"U"  
+        
+    def get_compact_qobj(self):
+        return self.U
+
+
+def create_controlled_unitary(controls, targets, U, control_value=1):
+    """
+    Create a controlled unitary gate.
+    
+    Parameters
+    ----------
+    controls : list
+        Control qubits
+    targets : list
+        Target qubits
+    U : Qobj
+        Unitary operator to apply on target qubits
+    control_value : int, optional
+        Control value (default: 1)
+        
+    Returns
+    -------
+    ControlledGate
+        The controlled unitary gate
+    """
+    return ControlledGate(
+        controls=controls,
+        targets=targets,
+        control_value=control_value,
+        target_gate=CustomGate,
+        U=U
+    )
+
+
+def _cphase_to_cnot(targets, controls, arg_value):
```

Comment:
We can import the `_cphase_to_cnot` directly from `qft.py`, even if it is not exposed to the namespace, no need to copy it. `from .algorithms.qft import ...` or something like this should work.